### DISCUSSION TITLE
SeleniumBase on RaspberryPi

### DISCUSSION THREAD

**Discussion by EdPS** [seleniumbase/SeleniumBase#2286] [Q&A] [answered]
Hi, I was testing SeleniumBase on my windows computer and it works perfectly. However, when I try to execute the same project on my raspberrypi4 it throws an error. 

`OSError: [Errno 8] Exec format error: '/home/usr/Projects/p1/venv/lib/python3.9/site-packages/seleniumbase/drivers/uc_driver'`

It seems that the uc_driver that is automatically downloaded is not compatible with ARM architecture. I tried searching for a uc_driver to download manually but could not find anything.

Any ideas on how to fix this? Many thanks.

**Comment by mdmintz** [comment, marked as answer]
Hello. `uc_driver` is generated from `chromedriver`, and the `chromedriver` downloads page only has one driver for Linux (Ubuntu):

<img width="680" alt="Screenshot 2023-11-15 at 3 24 32 PM" src="https://github.com/seleniumbase/SeleniumBase/assets/6788579/0d175783-e7c4-4e18-aae0-dab46f280051">

Due to that, there isn't official RaspberryPi support. However, there seems to be a lot of chat about it here:
https://github.com/search?q=repo%3Aultrafunkamsterdam%2Fundetected-chromedriver+Raspberry+Pi&type=issues

If you can find a chromedriver that supports it, then just rename it to `uc_driver` and drop it into the `seleniumbase/drivers` folder. Then it should work.

**Reply by EdPS** [reply]
Thank you so much. I wasn't sure whether uc_driver was the chromedriver renamed or something else. I'll give it a try. Kudos to your work. It has been very helpful.

**Reply by EdPS** [reply]
Hi @mdmintz. I installed the chromedriver and seleniumbase seems to be detecting it correctly. The program runs but it does not pass the bot detection from Cloudflare. The exact same code works on Windows though. 

When I run the code on Windows it displays this message in the terminal coming from seleniumbase: `DevTools listening on ws://127.0.0.1:59717/devtools/browser/74412774-a827-4f07-8ed9-ff522a77ed72`

That message does not show up un the raspberry pi. Any tips on how I could debug what is happening.?

Many thanks

**Reply by mdmintz** [reply]
For info on the Devtools message, see https://github.com/seleniumbase/SeleniumBase/issues/2292#issuecomment-1815255098

**Comment by EdPS** [comment]
Sorry I didn't explain myself crealy. The problem is not the message. The problem is that o Windows the anti-detection works but on the RaspberryPi does not. Exact same code running on both. I mentioned about the message because it shows on Windows but not on the RaspberryPi so I thought it might be a clue on something that runs differently on both architectures.

**Reply by mdmintz** [reply]
Until the chrome-for-testing downloads page has a unique driver for Raspberry Pi, the current Linux driver is only good enough for Ubuntu for UC Mode, it seems.

However, it’s also possible that this is all about the User Agent displayed when running on Raspberry Pi. You’ll need to set that yourself if the default one for Raspberry Pi gives it away that you’re using automation.

**Reply by EdPS** [reply]
Just to let you know I changed the user agent and that didn't fix it so it must be related to the driver or something else. Thank you for your prompt answers.

**Comment by lordlflm** [comment]
Hello, I found that there are arm64 drivers like found on https://github.com/electron/electron/releases .

I haven't read lot of UC/SeleniumBase code, but if my understanding is correct, UC applies a patch on the normal chromedriver to make it undetected.

A workaround was found : https://github.com/ultrafunkamsterdam/undetected-chromedriver/issues/917 , but Im not sure if it's up to date (have to test), because at first glance, it looks like UC code changed since the workaround was posted.

If someone knows how to patch chromdriver as of today, please let it know here :)
